File: introduction/tutorial/using-intervals-to-move-the-panda.cxx
#include "pandaFramework.h"
#include "pandaSystem.h"

#include "cIntervalManager.h"
#include "cLerpNodePathInterval.h"
#include "cMetaInterval.h"

int main(int argc, char *argv[]) {
  // Open a new window framework and set the title
  PandaFramework framework;
  framework.open_framework(argc, argv);
  framework.set_window_title("My Panda3D Window");

  // Open the window
  WindowFramework *window = framework.open_window();
  NodePath camera = window->get_camera_group(); // Get the camera and store it

  // Load the environment model
  NodePath scene = window->load_model(framework.get_models(), "models/environment");
  scene.reparent_to(window->get_render());
  scene.set_scale(0.25, 0.25, 0.25);
  scene.set_pos(-8, 42, 0);

  // Load our panda
  NodePath panda = window->load_model(framework.get_models(), "models/panda-model");
  panda.set_scale(0.005);
  panda.reparent_to(window->get_render());

  // Load the walk animation
  window->load_model(panda, "models/panda-walk4");
  window->loop_animations(0);

  // Create the lerp intervals needed to walk back and forth
  PT(CLerpNodePathInterval) pandaPosInterval1, pandaPosInterval2,
    pandaHprInterval1, pandaHprInterval2;
  pandaPosInterval1 = new CLerpNodePathInterval("pandaPosInterval1",
    13.0, CLerpInterval::BT_no_blend,
    true, false, panda, NodePath());
  pandaPosInterval1->set_start_pos(LPoint3(0, 10, 0));
  pandaPosInterval1->set_end_pos(LPoint3(0, -10, 0));

  pandaPosInterval2 = new CLerpNodePathInterval("pandaPosInterval2",
    13.0, CLerpInterval::BT_no_blend,
    true, false, panda, NodePath());
  pandaPosInterval2->set_start_pos(LPoint3(0, -10, 0));
  pandaPosInterval2->set_end_pos(LPoint3(0, 10, 0));

  pandaHprInterval1 = new CLerpNodePathInterval("pandaHprInterval1", 3.0,
    CLerpInterval::BT_no_blend,
    true, false, panda, NodePath());
  pandaHprInterval1->set_start_hpr(LPoint3(0, 0, 0));
  pandaHprInterval1->set_end_hpr(LPoint3(180, 0, 0));

  pandaHprInterval2 = new CLerpNodePathInterval("pandaHprInterval2", 3.0,
    CLerpInterval::BT_no_blend,
    true, false, panda, NodePath());
  pandaHprInterval2->set_start_hpr(LPoint3(180, 0, 0));
  pandaHprInterval2->set_end_hpr(LPoint3(0, 0, 0));

  // Create and play the sequence that coordinates the intervals
  PT(CMetaInterval) pandaPace;
  pandaPace = new CMetaInterval("pandaPace");
  pandaPace->add_c_interval(pandaPosInterval1, 0,
    CMetaInterval::RS_previous_end);
  pandaPace->add_c_interval(pandaHprInterval1, 0,
    CMetaInterval::RS_previous_end);
  pandaPace->add_c_interval(pandaPosInterval2, 0,
    CMetaInterval::RS_previous_end);
  pandaPace->add_c_interval(pandaHprInterval2, 0,
    CMetaInterval::RS_previous_end);
  pandaPace->loop();

  // Add our task, which can be any function or lambda that returns DoneStatus.
  framework.get_task_mgr().add("SpinCameraTask", [=](AsyncTask *task) mutable {
    // Calculate the new position and orientation (inefficient - change me!)
    double angledegrees = task->get_elapsed_time() * 6.0;
    double angleradians = angledegrees * (3.14 / 180.0);
    camera.set_pos(20 * sin(angleradians), -20.0 * cos(angleradians), 3);
    camera.set_hpr(angledegrees, 0, 0);

    // Tell the task manager to continue this task the next frame.
    return AsyncTask::DS_cont;
  });

  // Add another task to step the interval manager.
  framework.get_task_mgr().add("intervals", [](AsyncTask *task) {
    CIntervalManager::get_global_ptr()->step();
    return AsyncTask::DS_cont;
  });

  // Run Panda's main loop until the user closes the window.
  framework.main_loop();
  return 0;
}

File: introduction/tutorial/using-intervals-to-move-the-panda.py
from math import pi, sin, cos

from direct.showbase.ShowBase import ShowBase
from direct.task import Task
from direct.actor.Actor import Actor
from direct.interval.IntervalGlobal import Sequence
from panda3d.core import Point3


class MyApp(ShowBase):
    def __init__(self):
        ShowBase.__init__(self)

        # Disable the camera trackball controls.
        self.disable_mouse()

        # Load the environment model.
        self.scene = self.loader.load_model("models/environment")
        # Reparent the model to render.
        self.scene.reparent_to(self.render)
        # Apply scale and position transforms on the model.
        self.scene.set_scale(0.25, 0.25, 0.25)
        self.scene.set_pos(-8, 42, 0)

        # Add the spin_camera_task procedure to the task manager.
        self.taskMgr.add(self.spin_camera_task)

        # Load and transform the panda actor.
        self.panda_actor = Actor("models/panda-model",
                                {"walk": "models/panda-walk4"})
        self.panda_actor.set_scale(0.005, 0.005, 0.005)
        self.panda_actor.reparent_to(self.render)
        # Loop its animation.
        self.panda_actor.loop("walk")

        # Create the four lerp intervals needed for the panda to
        # walk back and forth.
        pos_ival1 = self.panda_actor.posInterval(13,
                                                 Point3(0, -10, 0),
                                                 startPos=Point3(0, 10, 0))
        pos_ival2 = self.panda_actor.posInterval(13,
                                                 Point3(0, 10, 0),
                                                 startPos=Point3(0, -10, 0))
        hpr_ival1 = self.panda_actor.hprInterval(3,
                                                 Point3(180, 0, 0),
                                                 startHpr=Point3(0, 0, 0))
        hpr_ival2 = self.panda_actor.hprInterval(3,
                                                 Point3(0, 0, 0),
                                                 startHpr=Point3(180, 0, 0))

        # Create and play the sequence that coordinates the intervals.
        self.pandaPace = Sequence(pos_ival1, hpr_ival1,
                                  pos_ival2, hpr_ival2,
                                  name="pandaPace")
        self.pandaPace.loop()

    # Define a procedure to move the camera.
    def spin_camera_task(self, task):
        angle_degrees = task.time * 6.0
        angle_radians = angle_degrees * (pi / 180.0)
        self.camera.set_pos(20 * sin(angle_radians), -20 * cos(angle_radians), 3)
        self.camera.set_hpr(angle_degrees, 0, 0)
        return Task.cont


app = MyApp()
app.run()
